README.md
# SODAQ Core for SAMD21 CPU

This repository contains the source code and configuration files of the SODAQ Core
for Atmel's SAMD21 processor (used on the SODAQ Autonomo board).

## Installation on Arduino IDE

This core is available as a package in the Arduino IDE cores manager.
You'll first have to add our SODAQ configuration.  Start the IDE and go to
File > Preferences > Additional Boards Manager URLs.  Add the following URL
`http://downloads.sodaq.net/package_sodaq_index.json`.  You may have to restart
the IDE.

Open the "Boards Manager" (Tools -> Board -> Boards Manager) and install the
package called:

"SODAQ SAMD Boards"

## Support

There is a dedicated section of the Arduino Forum for general discussion and project assistance:

http://forum.arduino.cc/index.php?board=98.0

## Bugs or Issues

If you find a bug you can submit an issue here on github:

https://github.com/arduino/SodaqCore-samd/issues

SodaqCore-samd is a clone of ArduinoCore-samd, but we created a branch "autonomo"
to add all the specific details of the SODAQ board.

Before posting a new issue, please check if the same problem has been already reported by someone else
to avoid duplicates.

## Contributions

Contributions are always welcome. The preferred way to receive code cotribution is by submitting a 
Pull Request on github.

## License and credits

This core has been developed by Arduino LLC in collaboration with Atmel.

```
  Copyright (c) 2015 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
```


CHANGELOG
SODAQ SAMD CORE v1.6.13-sodaq

This version is based of ArduionCore-samd 1.6.13, with the following
additions by SODAQ:

* change LORA_RESET to define
* Add a new KEEP section .sodaq_keep to be used for the signature
* mk-dist: Get version from platform.txt and check CORE_VERSION too
* SodaqOne: add pin 50 for PA19
* mk-dist: improve configurability
* boards: Correction name of bootloader for SODAQ ONE
* mk-dist: use a fixed top level dirname in the tar: SodaqCore-samd
* Exclude certain files from the distribution (such as *~, build, etc)
* Add __sodaq_firmware_state located at last 256 bytes of bootloader area
* Correct enum SercomUartSampleRate for ARITH x3
* Use an alternative baudrate "arith" method without need for float
* Correction of enum SAMPLE_RATE_ARITH_x8
* Correction of previous baudrate fix
* Optionally switch to non-fractional UART baudrate mode to support 300 baud
* Add package JSON files to this repo, and only have SAMD boards in it
* Correct typo in README
* Do not "self destruct" with 1200-bps touch but use bootloader "double tap"
* Add LORA_RESET
* Add __text_start__ to linker files, and align __data_end__ at 16 bytes
* Add bootloader for SODAQ ONE and rebuild the others too
* Remove all non-SODAQ variants and bootloaders
* sodaq_explorer: add specific bootloader for ExpLoRer
* bootloaders/autonomo: correct pin numbers for Serial RX/TX
* bootloaders/autonomo: add script to build all bootloaders
* bootloaders/autonomo: adapt changes after rebasing to ArduinoCore v1.6.11
* SodaqOne: Fix ACCEL_INT external int mappings
* sodaq_explorer: add LED_BUILTIN
* sodaq_explorer: add BUTTON
* BT_RESET
* Fixed a comment in sodaq_one/variant.cpp.
* Update copyright year in boards.txt
* Fixed an attribute on the LORA_RESET pin.
* Added LORA_RESET pin on PORTA7.
* Fixed/added some comments.
* Fixed the interrupt EIC numbers on pins D0/D1 (also RX/TX of Serial).
* Fixed the I2C entries for the Wire1 bus.
* Added SODAQ ONE with  ONEBase SPI sub-variant.
* White space cleanup in ExpLoRer files
* Added one extra pin to the ExpLoRer board. This pin will be the chip select of the flash IC
* Use lower case for file names, and upper case for (build.board) define
* New board files: sodaq ExpLoRer (microchip)
* bootloaders: add the bootloader for SODAQ Autonomo
* boards: rename LoRaONE Beta to SODAQ ONE Beta
* Comment and whitespace fixes.
* Renumbered the Digital pins (D0-D3, D6-D13).
* Renumbered Analog pins (A0-A3, A6-A13).
* Minor whitespace fix.
* Removed previously unnecessary PIN_A12, PIN_A13 definitions.
* Removed the hardware flow control from Serial1 on the Autonomo BETA.
* The CTS & RTS pins are now swapped for the Autonomo BETA subvariant.
* Changed the compiler directive to "AUTONOMO_BETA" in boards.txt
* Added "Autonomo BETA" subvariant to boards.txt
* Reformat the g_APinDescription tables in variants.cpp (white space only)
* Fix an issue with the PWM on the TIMER_ALT pins
* sodaq_one: fix PWM on D8 and D9, add BALT_VOLT
* Updated the Digital & Analog mappings removing D4/A4 & D5/A5.
* Moved SPI back to pins 3-6 of J1.
* Moved Serial1 back to Sercom2 (from 4).
* Removed I2C1
* Add initial support for SODAQ ONE with a copy from the LoRaONE Beta
* Add bootloader.tool and bootloader.file
* Moved SPI pins to A4-A7.
* Previous commit should have moved I2C1 to SERCOM2 not SERCOM0.
* Moved I2C1 from SERCOM4 to SERCOM0.
* Moved Serial1 from SERCOM2 to SERCOM4.
* Added second I2C bus (I2C1) on Sercom4 (SDA = D2/A2, SCL = D3/A3).
* Removed the setting of the digital IO pins to default INPUT.
* Rename LoRaONE to LoRaONE Beta
* Update CHANGELOG
* Added PWM capability to D8 D9.
* Add Autonomo sub-variant "... with SPI1"
* Update CHANGELOG
* Add ARDUINO_SAMD_VARIANT_COMPLIANCE 10603
* Added PWM to the RGB pins.
* Fixed Serial. Description required SERCOM_ALT.
* Fixed a spelling mistake for SWITCH_SENSE.
* Correction include guard of sodaq_wdt/variant.h
* There was a one off in the middle of g_APinDescription
* Rename board id for LoRaONE in boards.txt
* Increase SERIAL_BUFFER_SIZE to 129
* Renamed the openocd_scripts cfg file.
* Initial attempt at a pin mapping for the LoRaONE
* Add new board variant for WDT
* Correct chipname in debug script
* Add new variants "serial3" and "serial23"
* Add shell script to generate the distribution tar.bz2
* Add board variant with Serial2
* Add pins for SD_DFLASH and CS_SD to match the names on Autonomo rev4
* Avoid using void* in an arithmetic expression (in USBDeviceClass::send)
* Changed hardcoded A0 for DAC in analogWrite()
* Adding initialisation of TC6 & TC7
* Added the same clock divider rule as the G18
* Remove USB_HOST_ENABLE, we don't have the hardware to output USB power
* Make prototype of Serial_::begin consistent with implementation
* Add boards.txt, platform.txt and variants

SODAQ SAMD CORE v1.6.6-sodaq

SODAQ SAMD CORE v1.6.5-sodaq

SODAQ SAMD CORE v1.6.4-sodaq

This version is based of commit 90e234f, plus
(from the GIT shortlog)
* Undo the change in platform.txt for 1.6.6
* Add pins for SD_DFLASH and CS_SD to match the names on Autonomo rev4
* Avoid using void* in an arithmetic expression (in USBDeviceClass::send)
* Correct a few types in dtostrf.c to eliminate compilation warnings
* Changed to using floor(val) instea of iPart.
* This adds zeros after the decimal point if required.
* Added support for zero decimal places.
* A fix for the String class conversion of floats/doubles.
* Add strncpy_P and strncmp_P defines in pgmspace.h
* Add a a few missing enum values for ADC and TCx in WVariant.h
* Changed hardcoded A0 for DAC in analogWrite()
* Adding initialisation of TC6 & TC7
* Added the same clock divider rule as the G18
* Remove USB_HOST_ENABLE, we don't have the hardware to output USB power
* Make prototype of Serial_::begin consistent with implementation
* Add boards.txt, platform.txt and variants (for Autonomo)

SAMD CORE 1.6.13 2017.03.31

* Added SDU (SD card Updater) library.

cores/arduino/Arduino.h
/*
  Arduino.h - Main include file for the Arduino SDK
  Copyright (c) 2014 Arduino LLC.  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef Arduino_h
#define Arduino_h

#define CORE_VERSION "SodaqCore v1.6.13"

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef bool boolean;
typedef uint8_t byte;
typedef uint16_t word;

// some libraries and sketches depend on this AVR stuff,
// assuming Arduino.h or WProgram.h automatically includes it...
//
#include "avr/pgmspace.h"
#include "avr/interrupt.h"
#include "avr/io.h"

#include "binary.h"
#include "itoa.h"

#ifdef __cplusplus
extern "C"{
#endif // __cplusplus

// Include Atmel headers
#include "sam.h"

#include "wiring_constants.h"

#define clockCyclesPerMicrosecond() ( SystemCoreClock / 1000000L )
#define clockCyclesToMicroseconds(a) ( ((a) * 1000L) / (SystemCoreClock / 1000L) )
#define microsecondsToClockCycles(a) ( (a) * (SystemCoreClock / 1000000L) )

void yield( void ) ;

/* system functions */
int main( void );
void init( void );

/* sketch */
void setup( void ) ;
void loop( void ) ;

#include "WVariant.h"

#ifdef __cplusplus
} // extern "C"
#endif

// The following headers are for C++ only compilation
#ifdef __cplusplus
  #include "WCharacter.h"
  #include "WString.h"
  #include "Tone.h"
  #include "WMath.h"
  #include "HardwareSerial.h"
  #include "pulse.h"
#endif
#include "delay.h"
#ifdef __cplusplus
  #include "Uart.h"
#endif

// Include board variant
#include "variant.h"

#include "wiring.h"
#include "wiring_digital.h"
#include "wiring_analog.h"
#include "wiring_shift.h"
#include "WInterrupts.h"

// undefine stdlib's abs if encountered
#ifdef abs
#undef abs
#endif // abs

#define min(a,b) ((a)<(b)?(a):(b))
#define max(a,b) ((a)>(b)?(a):(b))
#define abs(x) ((x)>0?(x):-(x))
#define constrain(amt,low,high) ((amt)<(low)?(low):((amt)>(high)?(high):(amt)))
#define round(x)     ((x)>=0?(long)((x)+0.5):(long)((x)-0.5))
#define radians(deg) ((deg)*DEG_TO_RAD)
#define degrees(rad) ((rad)*RAD_TO_DEG)
#define sq(x) ((x)*(x))

#define interrupts() __enable_irq()
#define noInterrupts() __disable_irq()

#define lowByte(w) ((uint8_t) ((w) & 0xff))
#define highByte(w) ((uint8_t) ((w) >> 8))

#define bitRead(value, bit) (((value) >> (bit)) & 0x01)
#define bitSet(value, bit) ((value) |= (1UL << (bit)))
#define bitClear(value, bit) ((value) &= ~(1UL << (bit)))
#define bitWrite(value, bit, bitvalue) (bitvalue ? bitSet(value, bit) : bitClear(value, bit))

#define bit(b) (1UL << (b))

#if (ARDUINO_SAMD_VARIANT_COMPLIANCE >= 10606)
// Interrupts
#define digitalPinToInterrupt(P)   ( P )
#endif

// USB Device
#include "USB/USBDesc.h"
#include "USB/USBCore.h"
#include "USB/USBAPI.h"
#include "USB/USB_host.h"

#endif // Arduino_h
